primitives/message-lane/src/source_chain.rs

```diff
@@ -0,0 +1,103 @@
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity Bridges Common is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity Bridges Common.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Primitives of message lane module, that are used on the source chain.
+
+use crate::{LaneId, MessageNonce};
+
+use frame_support::Parameter;
+use sp_std::fmt::Debug;
+
+/// Target chain API. Used by source chain, to verify target chain proofs.
+///
+/// All implementations of this trait should only work with finalized data that
+/// can't change. Wrong implementation may lead to invalid lane states (i.e. lane
+/// that's stuck) and/or processing messages without paying fees.
+pub trait TargetHeaderChain<Payload> {
+	/// Error type.
+	type Error: Debug + Into<&'static str>;
+
+	/// Proof that messages have been received by target chain.
+	type MessagesReceivingProof: Parameter;
+	/// Proof that messages have been processed by target chain.
+	type MessagesProcessingProof: Parameter;
+
+	/// Verify message payload before we accept it.
+	///
+	/// **CAUTION**: this is very important function. Incorrect implementation may lead
+	/// to stucked lanes and/or relayers loses.
+	///
+	/// The proper implementation must ensure that the delivery-transaction with this
+	/// payload would (at least) be accepted into target chain transaction pool AND
+	/// eventually will be successfully 'mined'. The most obvious incorrect implementation
+	/// example would be implementation for BTC chain that accepts payloads larger than
+	/// 1MB. BTC nodes aren't accepting transactions that are larger than 1MB, so relayer
+	/// will be unable to craft valid transaction => this (and all subsequent) messages will
+	/// never be delivered.
+	fn verify_message(payload: &Payload) -> Result<(), Self::Error>;
+
+	/// Verify messages receiving proof and return lane && nonce of the latest recevied message.
+	fn verify_messages_receiving_proof(
+		proof: Self::MessagesReceivingProof,
+	) -> Result<(LaneId, MessageNonce), Self::Error>;
+	/// Verify messages processing proof and return lane && nonce of the latest processed message.
+	fn verify_messages_processing_proof(
+		proof: Self::MessagesProcessingProof,
+	) -> Result<(LaneId, MessageNonce), Self::Error>;
+}
+
+/// Lane message verifier.
+///
+/// Runtime developer may implement any additional validation logic over message-lane mechanism.
+/// E.g. if lanes should have some security (e.g. you can only accept Lane1 messages from
+/// Submitter1, Lane2 messages for those who has submitted first message to this lane, disable
+/// Lane3 until some block, ...), then it may be built using this verifier.
```

Comment:
I think we should mention that with current design it's important to make sure that the `fee` being passed is sufficient to deliver & dispatch the `payload`. I proposed in some other comment to actually return the minimal required `fee` instead of verifying it.

Reply:
Tips on selecting fee are mentioned in `MessageDeliveryAndDispatchPayment` documentation. As I said in other comment, imo we should not force `LaneMessageVerifier` to deal with `delivery_and_dispatch_fee` - it is like buyer-seller relations between sender and relayer. `LaneMessageVerifier` may just enforce some regulation over these relations - i.e. we know that minimal fee must be at least `N`, or no relayers will ever consider delivering this message. Knowledge of this `N` is out-of-scope of this PR.

Reply:
> - it is like buyer-seller relations between sender and relayer.

Well, I perceive this a bit differently. Choosing incorrect `fee` may result in the lane being stuck forever, so it's the lane's responsibility to make sure it attempts to ensure it's existence and usefullness. That said, the discussion should probably be part of a different PR.